torch/_inductor/pattern_matcher.py

```diff
@@ -552,6 +579,55 @@ def _match(self, node: torch.fx.Node, ctx: MatchContext):
         return m
 
 
+class PatternPrettyPrinter:
+    """
+    Serializes Patterns to executable python.
+    XXX: currently only used and tested for fuse attention patterns. May not cover
+    all patterns.
+    """
+
+    def __init__(self):
+        self.memoized_objs_names: Dict[PatternExpr, str] = {}
+        self.memoized_objs_pp: Dict[PatternExpr, str] = {}
+        self.tmp_counter = itertools.count(0)
+
+    @staticmethod
+    def run(obj: PatternExpr, output_name="output"):
+        """
+        Serializes obj to python code with obj written out to `output_name`
+        """
+
+        pp = PatternPrettyPrinter()
+        out_str = obj.pretty_print(pp=pp)
+
+        output = []
+        for key in pp.memoized_objs_names:
+            output.append(f"{pp.memoized_objs_names[key]} = {pp.memoized_objs_pp[key]}")
+
+        output.append(f"{output_name} = {out_str}")
+
+        return "\n".join(output)
+
+    def pretty_print(self, obj, level):
+        if isinstance(obj, _TargetArgsExpr):
```

Comment:
Why not just memoize all patterns?

Reply:
The memoization is needed for patterns that are used multiple times. IMO if we memoized everything it would make the patterns less readable since all CallFunctions would get written out to a temporary.

Example Before and After.

Before:

```
tmp_0 = CallFunction(aten.div.Tensor,
  CallFunction(aten.view.default,
    CallFunction(aten.bmm.default,
      CallFunction(aten.view.default,
        CallFunction(aten.expand.default,
          KeywordArg('query'),
          Ignored()
        ),
        Ignored()
      ),
      CallFunction(aten.view.default,
        CallFunction(aten.expand.default,
          CallFunction(aten.permute.default,
            KeywordArg('key'),
            Ignored()
          ),
          Ignored()
        ),
        Ignored()
      )
    ),
    Ignored()
  ),
  KeywordArg('inv_scale'),
  _users=2
)
tmp_1 = CallFunction(aten.exp.default,
  CallFunction(aten.sub.Tensor,
    tmp_0,
    CallFunction(aten.amax.default,
      tmp_0,
      Ignored(),
      True
    )
  ),
  _users=2
)
_sfdp_pattern_1_inference = CallFunction(aten.view.default,
  CallFunction(aten.bmm.default,
    CallFunction(aten.view.default,
      CallFunction(aten.expand.default,
        CallFunction(aten.div.Tensor,
          tmp_1,
          CallFunction(aten.sum.dim_IntList,
            tmp_1,
            Ignored(),
            True
          )
        ),
        Ignored()
      ),
      Ignored()
    ),
    CallFunction(aten.view.default,
      CallFunction(aten.expand.default,
        KeywordArg('value'),
        Ignored()
      ),
      Ignored()
    )
  ),
  Ignored(),
  _users=0
)
```

After:

```
tmp_0 = CallFunction(aten.expand.default,
  KeywordArg('query'),
  Ignored()
)
tmp_1 = CallFunction(aten.view.default,
  tmp_0,
  Ignored()
)
tmp_2 = CallFunction(aten.permute.default,
  KeywordArg('key'),
  Ignored()
)
tmp_3 = CallFunction(aten.expand.default,
  tmp_2,
  Ignored()
)
tmp_4 = CallFunction(aten.view.default,
  tmp_3,
  Ignored()
)
tmp_5 = CallFunction(aten.bmm.default,
  tmp_1,
  tmp_4
)
tmp_6 = CallFunction(aten.view.default,
  tmp_5,
  Ignored()
)
tmp_7 = CallFunction(aten.div.Tensor,
  tmp_6,
  KeywordArg('inv_scale'),
  _users=2
)
tmp_8 = CallFunction(aten.amax.default,
  tmp_7,
  Ignored(),
  True
)
tmp_9 = CallFunction(aten.sub.Tensor,
  tmp_7,
  tmp_8
)
tmp_10 = CallFunction(aten.exp.default,
  tmp_9,
  _users=2
)
tmp_11 = CallFunction(aten.sum.dim_IntList,
  tmp_10,
  Ignored(),
  True
)
tmp_12 = CallFunction(aten.div.Tensor,
  tmp_10,
  tmp_11
)
tmp_13 = CallFunction(aten.expand.default,
  tmp_12,
  Ignored()
)
tmp_14 = CallFunction(aten.view.default,
  tmp_13,
  Ignored()
)
tmp_15 = CallFunction(aten.expand.default,
  KeywordArg('value'),
  Ignored()
)
tmp_16 = CallFunction(aten.view.default,
  tmp_15,
  Ignored()
)
tmp_17 = CallFunction(aten.bmm.default,
  tmp_14,
  tmp_16
)
_sfdp_pattern_1_inference = CallFunction(aten.view.default,
  tmp_17,
  Ignored(),
  _users=0
)
```

I don't have a strong preference here and happy to do either.

Reply:
I guess we could use more semantic names then `tmp` and it would be more readable

Reply:
switched to using semantic name, looks good now